src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp

```diff
@@ -1072,14 +1072,21 @@ Block ArrowColumnToCHColumn::arrowSchemaToCHHeader(
         auto arrow_column = std::make_shared<arrow::ChunkedArray>(array_vector);
         std::unordered_map<std::string, DictionaryInfo> dict_infos;
         bool skipped = false;
-        bool allow_null_type = false;
+        bool allow_null_type = field->nullable();
         if (hint_header && hint_header->has(field->name()) && hint_header->getByName(field->name()).type->isNullable())
             allow_null_type = true;
+
         ColumnWithTypeAndName sample_column = readColumnFromArrowColumn(
             arrow_column, field->name(), format_name, false, dict_infos, allow_null_type, skip_columns_with_unsupported_types, skipped);
-        if (!skipped)
-            sample_columns.emplace_back(std::move(sample_column));
+        if (skipped)
+            continue;
+
+        if (allow_null_type && sample_column.type->canBeInsideNullable())
+            sample_column.type = makeNullable(sample_column.type);
```

Comment:
So, we use information about field nullability only for top-level fields, so it won't work for nested types like `Array(Nullable(...))`/`Tuple(Nullable(...), ...)`/`Map(KeyType, Nullable(...))`.

In `readColumnFromArrowColumn` we have a logic of making columns `Nullable` in 2 cases: first when arrow column contains nulls, second when type hint was provided and it's `Nullable`. In schema inference we don't have any of that. We can add one more bool flag `make_nullable` next to `is_nullable` in `readColumnFromArrowColumn` that will be third option of making column `Nullable`, this flag will be checked here (`arrow_column->null_count() || make_nullable || ...)`:
https://github.com/ClickHouse/ClickHouse/blob/0244e0851a503c948da1b0b6be4caec9cbd0951a/src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp#L721

And we can propagate this flag recursively when we process Map/Array/Struct arrow types.

Reply:
Also, we can remove `hint_header` and `ignore_case` arguments from `arrowSchemaToCHHeader` functions, we never actually pass them and use default values.

Reply:
After these changes we can stop using setting `schema_inference_make_columns_nullable` in Arrow and Paarquet formats, as `Nullable` types will be inferred automatically.

Reply:
Updated implementation.
For now it make sense to have `schema_inference_make_columns_nullable` enabled by default, because otherwise it can break backward compatibility and a lot of tests.
Maybe it make sense to disable it in separate pull request.